samples/nrf9160/gps/src/main.c

```diff
@@ -130,85 +133,97 @@ static int activate_lte(bool activate)
+		fix_timestamp = k_uptime_get();
+		break;
+
+	case NRF_MODEM_GNSS_EVT_NMEA:
+		nmea_data = k_malloc(sizeof(struct nrf_modem_gnss_nmea_data_frame));
+
+		retval = nrf_modem_gnss_read(nmea_data,
+					     sizeof(struct nrf_modem_gnss_nmea_data_frame),
+					     NRF_MODEM_GNSS_DATA_NMEA);
+		if (retval == 0) {
+			retval = k_msgq_put(&nmea_queue, &nmea_data, K_NO_WAIT);
 		} else {
-			printk("Could not init socket (err: %d)\n", gnss_fd);
-			return -1;
+			k_free(nmea_data);
 		}
+		break;
 
-		retval = nrf_setsockopt(gnss_fd,
-					NRF_SOL_GNSS,
-					NRF_SO_GNSS_FIX_RETRY,
-					&fix_retry,
-					sizeof(fix_retry));
-		if (retval != 0) {
-			printk("Failed to set fix retry value\n");
-			return -1;
+	case NRF_MODEM_GNSS_EVT_AGPS_REQ:
+#ifdef CONFIG_SUPL_CLIENT_LIB
+		retval = nrf_modem_gnss_read(&last_agps,
+					     sizeof(last_agps),
+					     NRF_MODEM_GNSS_DATA_AGPS_REQ);
+		if (retval == 0) {
+			agps_requested = true;
 		}
+#endif
+		break;
 
-		retval = nrf_setsockopt(gnss_fd,
-					NRF_SOL_GNSS,
-					NRF_SO_GNSS_FIX_INTERVAL,
-					&fix_interval,
-					sizeof(fix_interval));
-		if (retval != 0) {
-			printk("Failed to set fix interval value\n");
-			return -1;
-		}
+	case NRF_MODEM_GNSS_EVT_BLOCKED:
+		gnss_blocked = true;
+		break;
 
-		retval = nrf_setsockopt(gnss_fd,
-					NRF_SOL_GNSS,
-					NRF_SO_GNSS_NMEA_MASK,
-					&nmea_mask,
-					sizeof(nmea_mask));
-		if (retval != 0) {
-			printk("Failed to set nmea mask\n");
-			return -1;
-		}
+	case NRF_MODEM_GNSS_EVT_UNBLOCKED:
+		gnss_blocked = false;
+		break;
+
+	default:
+		break;
 	}
+}
+
+static int gnss_ctrl(uint32_t ctrl)
+{
+	int retval;
+
+	switch (ctrl) {
+	case GNSS_INIT_AND_START:
+		retval = nrf_modem_gnss_init();
+		retval |= nrf_modem_gnss_event_handler_set(gnss_event_handler);
+
+		retval |= nrf_modem_gnss_nmea_mask_set(NRF_MODEM_GNSS_NMEA_RMC_MASK |
+						       NRF_MODEM_GNSS_NMEA_GGA_MASK |
+						       NRF_MODEM_GNSS_NMEA_GLL_MASK |
+						       NRF_MODEM_GNSS_NMEA_GSA_MASK |
+						       NRF_MODEM_GNSS_NMEA_GSV_MASK);
+		retval |= nrf_modem_gnss_fix_retry_set(0);
+		retval |= nrf_modem_gnss_fix_interval_set(1);
 
-	if ((ctrl == GNSS_INIT_AND_START) ||
-	    (ctrl == GNSS_RESTART)) {
-		retval = nrf_setsockopt(gnss_fd,
-					NRF_SOL_GNSS,
-					NRF_SO_GNSS_START,
-					&delete_mask,
-					sizeof(delete_mask));
 		if (retval != 0) {
-			printk("Failed to start GPS\n");
+			printk("Failed to initialize GNSS interface");
 			return -1;
 		}
-	}
+		/* fall through */
+	case GNSS_RESTART:
+		if (nrf_modem_gnss_start() != 0) {
+			printk("Failed to start GNSS\n");
+			return -1;
+		}
+		break;
 
-	if (ctrl == GNSS_STOP) {
-		retval = nrf_setsockopt(gnss_fd,
-					NRF_SOL_GNSS,
-					NRF_SO_GNSS_STOP,
-					&delete_mask,
-					sizeof(delete_mask));
-		if (retval != 0) {
-			printk("Failed to stop GPS\n");
+	case GNSS_STOP:
+		if (nrf_modem_gnss_stop() != 0) {
+			printk("Failed to stop GNSS\n");
 			return -1;
 		}
+		break;
+
+	default:
+		return -1;
```

Comment:
Could print a log message here that the control command was not recognized

Reply:
Refactored the code as requested by @lemrey and removed the whole function.